### DISCUSSION TITLE
Issues with implementing public.users table

### DISCUSSION THREAD

**Discussion by ox-harris** [supabase/supabase#2962] [Questions] [answered]
In my attempt to create the public.users table and bind it to the auth.users table, I run into issues. With my SQL code below, I get an error when I try to invite a user by email from the Supabase UI. Error: "Inviting user failed: Database error saving new user".

The trigger:

```sql
CREATE OR REPLACE FUNCTION public.sp_create_user_at_public()
RETURNS TRIGGER language plpgsql as $$
BEGIN
    INSERT INTO public.users (id, full_name, avatar_url)
    VALUES (NEW.id, NEW.raw_user_meta_data->>'full_name', NEW.raw_user_meta_data->>'avatar_url');
    RETURN NEW;
END $$; 
DROP TRIGGER IF EXISTS create_user_at_public ON auth.users;
CREATE TRIGGER create_user_at_public AFTER INSERT ON auth.users FOR EACH ROW EXECUTE PROCEDURE public.sp_create_user_at_public();
```

The `public.users` table:

```sql
DROP TABLE IF EXISTS users CASCADE;
CREATE TABLE users (
    id UUID REFERENCES auth.users NOT NULL PRIMARY KEY,
    plan_id INT REFERENCES plans (id) ON DELETE RESTRICT,
    full_name text,
    avatar_url text,
    UNIQUE (id, plan_id)
); CREATE TRIGGER on_update_current_timestamp BEFORE UPDATE ON users FOR EACH ROW EXECUTE PROCEDURE sp_on_update_current_timestamp();
```

**Comment by silentworks** [comment]
You need to set the default value for `full_name` and `avatar_url` to null. Right now it's expecting these to have actual values and none is being passed when using the invite feature inside the Supabase Dashboard.

**Reply by ox-harris** [reply]
Thanks for your swift response.

Same error with default values in place, tho.

**Reply by silentworks** [reply]
What does your `CREATE TABLE users` statement look like now?

**Reply by ox-harris** [reply]
```sql
CREATE TABLE users (
    id UUID REFERENCES auth.users NOT NULL PRIMARY KEY,
    plan_id INT REFERENCES plans (id) ON DELETE RESTRICT,
    full_name text DEFAULT NULL,
    avatar_url text DEFAULT NULL,
    UNIQUE (id, plan_id)
)
```

**Comment by w3b6x9** [comment, marked as answer]
@ox-harris this might be a database permissions issue.

The function `public.sp_create_user_at_public()` when triggered outside of Supabase Dashboard's SQL editor will fail b/c the SQL editor environment has super admin privileges and the function contains inserting to `public.users` that has a reference to `auth.users` (limited permissions).

You can test by creating the same trigger function with `SECURITY DEFINER`:

```
CREATE OR REPLACE FUNCTION public.sp_create_user_at_public()
RETURNS TRIGGER language plpgsql as $$
BEGIN
    INSERT INTO public.users (id, full_name, avatar_url)
    VALUES (NEW.id, NEW.raw_user_meta_data->>'full_name', NEW.raw_user_meta_data->>'avatar_url');
    RETURN NEW;
END $$
SECURITY DEFINER;
```

Now, no matter who triggers that function it'll run with the permissions of the role who created it (creating this in the SQL editor will grant super admin privileges). But please be careful when defining functions this way. Consult the `Writing
SECURITY DEFINER Functions Safely` section at https://www.postgresql.org/docs/13/sql-createfunction.html for best security practices.

**Reply by ox-harris** [reply]
This killed it indeed @w3b6x9 . Much appreciated.
